lib/new_relic/agent/log_event_aggregator.rb

```diff
@@ -0,0 +1,209 @@
+        })
+
+        [
+          {
+            PrioritySampledBuffer::PRIORITY_KEY => priority
+          },
+          event
+        ]
+      end
+
+      # Because our transmission format (MELT) is different than historical
+      # agent payloads, extract the munging here to keep the service focused
+      # on the general harvest + transmit instead of the format.
+      #
+      # Payload shape matches the publicly documented MELT format.
+      # https://docs.newrelic.com/docs/logs/log-api/introduction-log-api
+      #
+      # We have to keep the aggregated payloads in a separate shape, though, to
+      # work with the priority sampling buffers
+      def self.payload_to_melt_format(data)
+        common_attributes = LinkingMetadata.append_service_linking_metadata({})
+
+        # To save on unnecessary data transmission, trim the name and type
+        # that were sent by classic logs-in-context
+        common_attributes.delete(ENTITY_NAME_KEY)
+        common_attributes.delete(ENTITY_TYPE_KEY)
+
+        _, items = data
+        payload = [{
+          common: { attributes: common_attributes },
+          logs: items.map(&:last)
+        }]
+
+        return [payload, items.size]
+      end
+
+      def harvest!
+        record_customer_metrics()
+        super
+      end
+
+      def reset!
+        @counter_lock.synchronize do
+          @seen = 0
+          @seen_by_severity.clear
+        end
+        super
+      end
+
+      private
+
+      # We record once-per-connect metrics for enabled/disabled state at the
+      # point we consider the configuration stable (i.e. once we've gotten SSC)
+      def register_for_done_configuring(events)
+        events.subscribe(:server_source_configuration_added) do
+          @high_security = NewRelic::Agent.config[:high_security]
+
+          record_configuration_metric(METRICS_SUPPORTABILITY_FORMAT, METRICS_ENABLED_KEY)
+          record_configuration_metric(FORWARDING_SUPPORTABILITY_FORMAT, FORWARDING_ENABLED_KEY)
+          record_configuration_metric(DECORATING_SUPPORTABILITY_FORMAT, DECORATING_ENABLED_KEY)
+        end
+      end
+
+      def record_configuration_metric(format, key)
+          state = NewRelic::Agent.config[key]
+          label = state ? "enabled" : "disabled"
+          NewRelic::Agent.increment_metric(format % label)
+      end
+
+      def after_harvest metadata
+        dropped_count = metadata[:seen] - metadata[:captured]
+        note_dropped_events(metadata[:seen], dropped_count)
+        record_supportability_metrics(metadata[:seen], metadata[:captured], dropped_count)
+      end
+
+      # To avoid paying the cost of metric recording on every line, we hold
+      # these until harvest before recording them
+      def record_customer_metrics
+        @counter_lock.synchronize do
+          return unless @seen > 0
+
+          NewRelic::Agent.increment_metric(LINES, @seen)
+          @seen_by_severity.each do |(severity, count)|
```

Comment:
This syntax is new to me! Why are parenthesis used within the iterator pipes?

Reply:
So that's array destructuring. The `each` on `@seen_by_severity` returns an array of `[key, value]` arrays, and this will put those positional elements into the variables listed in the `()`.

Example pry session (with bonus "remainder" gathering we aren't bothering with here.)

```
[5] pry(main)> pr = Proc.new do |(first, second, *rest)| p first; p second; p rest; end
=> #<Proc:0x00007fcbda141060 (pry):5>

[6] pry(main)> pr.call([1,2,3,4])
1
2
[3, 4]

[7] pry(main)> pr.call([1,2,3])
1
2
[3]

[8] pry(main)> pr.call([1,2])
1
2
[]
```